applications/generate/philadelphia/orchestra.py

```diff
@@ -0,0 +1,129 @@
+#
+# Copyright 2020 Philadelphia authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import xml.etree.ElementTree
+import collections
+
+from . import model
+
+
+NS = {
+    'fixr': 'http://fixprotocol.io/2020/orchestra/repository',
+}
+
+
+def read_messages(filename):
+    def message(elem):
+        name = elem.get('name')
+        msg_type = elem.get('msgType')
+        return model.Message(name, msg_type)
+    tree = xml.etree.ElementTree.parse(filename)
+    return [message(elem) for elem in tree.findall('.//fixr:message', NS)]
+
+
+def read_fields(filename):
+    tree = xml.etree.ElementTree.parse(filename)
+    fields = _read_fields(tree)
+    code_sets = _read_code_sets(tree)
+    return sorted([_field(field, code_sets) for field in fields],
+                  key=lambda field: int(field.tag))
+
+
+def _field(field, code_sets):
+    tag = field.id_
+    name = field.name
+    code_set = code_sets.get(tag)
+    type_ = _type(field, code_set)
+    has_values = code_set and _has_values(code_set)
+    values = [_value(tag, code) for code in code_set.codes] if has_values else []
+    return model.Field(tag, name, _TYPES.get(type_, type_), values)
+
+
+def _type(field, code_set):
+    if not code_set:
+        return field.type_
+    if code_set.type_ == 'char' and max(len(code.value) for code in code_set.codes) > 1:
+        return 'String'
+    return code_set.type_
+
+
+_NO_VALUES = [
+    '35',
+    '41237',
+]
+
+
+def _has_values(code_set):
+    return not code_set.id_ in _NO_VALUES and not code_set.type_ == 'Boolean'
+
+
+_NAME_REPLACEMENTS = {
+    ('1779', '1'): 'Int',
+    ('1779', '6'): 'Float',
+    ('1779', '12'): 'Char',
+}
```

Comment:
I have opened [an issue on the FIX Trading Community repository](https://github.com/FIXTradingCommunity/orchestrations/issues/15) regarding the handling of these values.

Reply:
This feels like the best way to handle these for now.